fix(wool): use platform-specific ColorUtils for wool matching

TTtie opened the pull request:
On 1.21 servers, CraftItemStack#getData() returns unsubclassed MaterialData for wool item stacks, hence we cannot check for `instanceof Wool` there. This patch changes the appropriate code to a manual comparison, enabling pickup and proximity tracking to work regardless of whether the material data is correctly subclassed.

TTtie committed:
fix(MonumentWoolFactory): manually check for item type and color

On 1.21 servers, CraftItemStack#getData() returns unsubclassed MaterialData for wool item stacks, hence we cannot check for `instanceof Wool` there. This patch changes the appropriate code to a manual comparison, enabling pickup and proximity tracking to work regardless of whether the material data is correctly subclassed.

Signed-off-by: TTtie <me@tttie.cz>

cswhite2000 commented on core/src/main/java/tc/oc/pgm/wool/MonumentWoolFactory.java:
```diff
@@ -123,7 +123,10 @@ public boolean isObjectiveWool(ItemStack stack) {
-    return material instanceof Wool && ((Wool) material).getColor() == this.color;
+    // On 1.21, CraftItemStack#getData() on wool item stacks returns unsubclassed MaterialData,
+    // hence we're comparing the item type and color manually
+    return material.getItemType() == Materials.WOOL
```
For 1.13+ the Wool material is always in the format `COLOR_WOOL`. This won't work for modern versions.

Pablete1234 replied:
it will work because we are manually defining the wools with legacy wool, but it's not something we should do anyways

TTtie committed:
refactor(wool): use platform-specific ColorUtils to determine color correctness

Signed-off-by: TTtie <me@tttie.cz>

cswhite2000 commented on core/src/main/java/tc/oc/pgm/wool/WoolMatchModule.java:
```diff
@@ -83,7 +80,7 @@ public Multimap<Team, MonumentWool> getWools() {
 
   private boolean isObjectiveWool(ItemStack stack) {
-    if (WOOL.matches(stack.getType())) {
+    if (WoolModule.WOOL.matches(stack.getType())) {
```
nitpick: static importing this would remove the diff

cswhite2000 commented on core/src/main/java/tc/oc/pgm/wool/WoolModule.java:
```diff
@@ -29,11 +29,13 @@
 import tc.oc.pgm.util.xml.XMLUtils;
 
 public class WoolModule implements MapModule<WoolMatchModule> {
+  static final MaterialMatcher WOOL = MaterialMatcher.of(m -> m.name().endsWith("WOOL"));
```
nitpick: Style: prefer to have an explicit access modifier over an implicit one

TTtie replied:
This is intentionally package-private, how do I document this? Or should I avoid package-private access level and bump this to `public`?

cswhite2000 replied:
If you have a reason for it to be packaged private over protected than that is fine. In any case this isn't a sufficient issue to block merging

cswhite2000 commented on platform/platform-sportpaper/src/main/java/tc/oc/pgm/platform/sportpaper/material/SpColorUtils.java:
```diff
@@ -71,6 +71,11 @@ public boolean isColor(BlockState block, DyeColor color) {
 
+  @Override
+  public boolean isColor(ItemStack stack, DyeColor color) {
+    return color.getWoolData() == stack.getDurability();
```
If this is called on other things with data / durability this may return true even if color is not relevant.
Not sure if this is currently an issue based off of current usage of the method, but it seems a candidate for easily introducing bugs if kept as is.
If you double check current usage and see no issues, then I'm fine with keeping it as is for this pr.

TTtie replied:
Currently, this method is used only here: (the mentioned method is called by other 4 callers, but the preceding wool query filters that away)

https://github.com/PGMDev/PGM/blob/0065f1527c2e85683f66d6d2d2d6f8c0b69b6aa1/core/src/main/java/tc/oc/pgm/wool/MonumentWoolFactory.java#L123-L125

As for this method being called for un-colorable items, I agree this might open a possibility for other bugs, but none of the other methods in legacy ColorUtils check for whether the block can be colored either – fixing this should go in another PR as it would be out of scope of this PR and #1363 for which this PR is set to be merged into.

cswhite2000 reviewed (commented):
nothing major, few minor things

TTtie committed:
refactor(wool): statically import WoolModule.WOOL

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 commented on platform/platform-modern/src/main/java/tc/oc/pgm/platform/modern/material/ModernColorUtils.java:
```diff
@@ -103,9 +103,18 @@ public void setColor(Block block, DyeColor color) {
     block.setType(setColor(block.getType(), color));
   }
 
+  private boolean setAndCompare(Material mat, DyeColor color) {
+    return setColor(mat, color) == mat;
+  }
+
```
```suggestion
```

i'd inline it, it's such a trivial thing

Pablete1234 commented on util/src/main/java/tc/oc/pgm/util/material/ColorUtils.java:
```diff
@@ -29,6 +29,8 @@ public interface ColorUtils {
 
   boolean isColor(BlockState block, DyeColor color);
 
+  boolean isColor(ItemStack stack, DyeColor color);
```
This should probably just use pgm's MaterialData as param instead, so it can work for both blockstte and itemstacks: `MaterialData.of(blockstate)` or `MaterialData.of(stack)`

TTtie committed:
refactor(ColorUtils): unify color checking

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
nit(WoolModule): bump WoolModule.WOOL access to protected

Per https://github.com/PGMDev/PGM/pull/1364#discussion_r1672800632 explicit access modifiers are preferred and there is no good reason to ban possible subclasses of WoolModule from accessing the material matcher

Signed-off-by: TTtie <me@tttie.cz>
